woocommerce/woocommerce
# [COT] Log errors during the posts to orders table migrations

### All Submissions:


### Changes proposed in this Pull Request:

Implement logging for errors that happen during the execution of `PostsToOrdersMigrationController::migrate_orders`. The following types of errors are logged:

1. Unhandled exceptions.
2. Database errors caused by `$wpdb->query` and `$wpdb->get_results`.
3. Not enough rows processed after an insert or update operation.
4. Validation errors on the data to be migrated.

Errors are logged using the standard WooCommerce logger, using `posts-to-orders-migration` as the source name.

Closes #32664.

### How to test the changes in this Pull Request:

First, enable the feature by adding the following code snippet somewhere, for example at the end of `woocommerce.php`:

```php
wc_get_container()
   ->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController')
   ->show_feature();
```

Then force errors of each type, then go to WooCommerce - Status - Logs and verify that the errors are properly logged.

One way to force errors is to to temporarily modify the code as follows (don't try more than one at the same time tas it would get confusing):

1. In `TableMigrator::process_migration_batch_for_ids`, right before the first statement of the `try` block, add this: `throw new \Exception("Ouch!!");`
2. In `TableMigrator::db_query`, right before the `return` statement, add this: `$this->add_error('Gasp!!');` (also in `db_get_results`)
3. In `TableMigrator::maybe_add_insert_or_update_mismatch_error`, right after `$count` is calculated increase it by one.
4. In `MetaToCustomTableMigrator::validate_data`, add this right before the first line of the method: `return new \WP_Error('Argh!!');`

Feel free to use other (probably more proper) ways to force errors. For example you could temporarily remove or rename the `wp_wc_orders` table.

### Other information:


### FOR PR REVIEWER ONLY:
